bootstrap/src/sun/nio/ch/lincheck/Types.java

```diff
@@ -0,0 +1,311 @@
+        }
+
+        @Override
+        public boolean equals(java.lang.Object obj) {
+            if (this == obj) return true;
+            if (!(obj instanceof MethodSignature)) return false;
+
+            MethodSignature other = (MethodSignature) obj;
+            return (
+                name.equals(other.name) &&
+                methodType.equals(other.methodType)
+            );
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(name, methodType);
+        }
+    }
+
+    // ------ Types ---------------------- //
+    public abstract static class Type {
+        public static final class Void extends Type {
+            private Void() {}
+            static class Holder {
+                static final Void INSTANCE = new Void();
+            }
+            public static Void get() {
+                return Holder.INSTANCE;
+            }
+        }
+    }
+
+    public abstract static class ArgumentType extends Type {
+        public static final class Object extends ArgumentType {
+            private final String className;
+
+            public Object(String className) {
+                this.className = className;
+            }
+
+            public String getClassName() {
+                return className;
+            }
+
+            @Override
+            public boolean equals(java.lang.Object obj) {
+                if (this == obj) return true;
+                if (!(obj instanceof ArgumentType.Object)) return false;
+
+                ArgumentType.Object other = (ArgumentType.Object) obj;
+                return className.equals(other.className);
+            }
+
+            @Override
+            public int hashCode() {
+                return Objects.hash(className);
+            }
+        }
+        public static final class Array extends ArgumentType {
+            private final ArgumentType type;
+
+            public Array(ArgumentType type) {
+                this.type = type;
+            }
+
+            public ArgumentType getType() {
+                return type;
+            }
+
+            @Override
+            public boolean equals(java.lang.Object obj) {
+                if (this == obj) return true;
+                if (!(obj instanceof ArgumentType.Array)) return false;
+
+                ArgumentType.Array other = (ArgumentType.Array) obj;
+                return type.equals(other.type);
+            }
+
+            @Override
+            public int hashCode() {
+                return Objects.hash(type);
+            }
+        }
+
+        public abstract static class Primitive extends ArgumentType {
+            public static final class Int extends Primitive {
+                private Int() {}
+                private static class Holder {
+                    static final Int INSTANCE = new Int();
+                }
+                public static Int get() {
+                    return Int.Holder.INSTANCE;
+                }
+            }
```

Comment:
Why not just: 
```java
public static final class Int extends Primitive {}
public static final Int int = new Int();
```

Same for other primitive types.

Reply:
Idea highlight a warning here: `Referencing subclass Int from superclass Primitive initializer might lead to class loading deadlock`.

Related to https://stackoverflow.com/questions/50021182/java-static-initializers-referring-to-subclasses-avoid-class-loading-deadlock

Reply:
Can move object initialization:
```java
public static final Int int = new Int();
```
out of enclosing parent class?
Would it solve the issue?

Reply:
yeah, since I have flatten the nested classes, now I can do that